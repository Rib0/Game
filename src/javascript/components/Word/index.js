import React, { PureComponent } from 'react';

class Word extends PureComponent {
    state = {
        word: [
            {
                letter: '',
                complete: false,
            },
        ],
    };

    componentDidUpdate() {
        const { word } = this.props;
        const stateWord = word.split('').map(letter => ({ letter, complete: false }));

        this.setState({
            word: stateWord,
        });
    }

    onKeyDown = () => {};

    render() {
        const { word } = this.state;

        return <div />;
    }
}

export default Word;
