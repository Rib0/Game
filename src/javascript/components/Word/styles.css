.word {
    display: flex;
    justify-content: center;
}

.letter {
    position: relative;
    font-size: 90px;
    font-weight: bold;
    color: var(--colorWhite);
    overflow: hidden;

    &.complete > .inner {
        width: 100%;
    }
}

.inner {
    width: 0;
    position: absolute;
    top: 0;
    left: 0;
    transition: width .2s;
    color: var(--colorOrange);
    overflow: hidden;
    white-space: nowrap;
}