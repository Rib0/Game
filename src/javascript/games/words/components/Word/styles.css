@keyframes missedAnimation {
    0% {
        color: red;
    }

    100% {
        color: auto
    }
}

.word {
    display: flex;
    justify-content: center;
}

.letter {
    position: relative;
    font-size: 90px;
    font-weight: bold;
    color: var(--colorWhite);
    overflow: hidden;

    &.missed {
        animation: missedAnimation .1s steps(1);
    }

    &.complete > .inner {
        width: 100%;
    }
}

.inner {
    width: 0;
    position: absolute;
    top: 0;
    left: 0;
    transition-property: width;
    transition-timing-function: linear;
    color: var(--colorOrange);
    overflow: hidden;
    white-space: nowrap;
}
