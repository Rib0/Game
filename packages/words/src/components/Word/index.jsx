import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import styles from './styles.css';

const cx = classNames.bind(styles);

class Word extends PureComponent {
    state = {
        transitionDuration: '.2s',
        missedIndex: null,
    };

    componentDidMount() {
        document.body.addEventListener('keydown', this.onKeyDown);
    }

    resetTransition() {
        this.setState({
            transitionDuration: '0s',
        });
        setTimeout(() => {
            this.setState({
                transitionDuration: '.2s',
            });
        }, 0);
    }

    componentDidUpdate(prevProps) {
        const { wordArray, getRightWord } = this.props;

        if (prevProps.wordArray !== wordArray) {
            const isComplete = wordArray.every(word => word.complete);
            if (isComplete) {
                getRightWord();
                this.resetTransition();
            }
        }
    }

    handleMiss() {
        const { wordArray } = this.props;

        const missedLetterIndex = wordArray.findIndex(word => !word.complete);
        this.setState({ missedIndex: missedLetterIndex });
        setTimeout(() => {
            this.setState({ missedIndex: null });
        }, 100);
    }

    onKeyDown = event => {
        const { key } = event;
        const { wordArray, getRightLetter, resetWord } = this.props;

        const currentLetter = wordArray.find(word => !word.complete);

        const { letter } = currentLetter;

        if (letter === key) {
            getRightLetter();
        } else {
            this.handleMiss();
            resetWord();
            this.resetTransition();
        }
    };

    componentWillUnmount() {
        document.body.removeEventListener('keydown', this.onKeyDown);
    }

    render() {
        const { wordArray } = this.props;
        const { transitionDuration, missedIndex } = this.state;

        return (
            <div className={styles.word}>
                {wordArray.map(({ letter, complete }, index) => (
                    <span
                        key={index}
                        className={cx('letter', { complete, missed: missedIndex === index })}
                    >
                        {letter}
                        <div style={{ transitionDuration }} className={styles.inner}>
                            {letter}
                        </div>
                    </span>
                ))}
            </div>
        );
    }
}

Word.propTypes = {
    wordArray: PropTypes.arrayOf(
        PropTypes.shape({
            letter: PropTypes.string,
            complete: PropTypes.bool,
            missed: PropTypes.bool,
        })
    ),
    getRightWord: PropTypes.func,
    getRightLetter: PropTypes.func,
    resetWord: PropTypes.func,
};

export default Word;
