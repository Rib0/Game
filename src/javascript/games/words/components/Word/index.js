import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import styles from './styles.css';

const cx = classNames.bind(styles);

class Word extends PureComponent {
    state = {
        transitionDuration: '.2s',
    };

    componentDidMount() {
        document.body.addEventListener('keydown', this.onKeyDown);
    }

    resetTransition() {
        this.setState({
            transitionDuration: '0s',
        });
        setTimeout(() => {
            this.setState({
                transitionDuration: '.2s',
            });
        }, 0);
    }

    componentDidUpdate(prevProps) {
        const { wordArray, getRightWord } = this.props;

        if (prevProps.wordArray !== wordArray) {
            const isComplete = wordArray.every(word => word.complete);
            isComplete && (getRightWord(), this.resetTransition());
        }
    }

    onKeyDown = event => {
        const { key } = event;
        const { wordArray, getRightLetter, resetWord } = this.props;
        const currentLetter = wordArray.find(word => !word.complete);

        const { letter } = currentLetter;

        letter === key ? getRightLetter() : (resetWord(), this.resetTransition());
    };

    componentWillUnmount() {
        document.body.removeEventListener('keydown', this.onKeyDown);
    }

    render() {
        const { wordArray } = this.props;
        const { transitionDuration } = this.state;

        return (
            <div className={styles.word}>
                {wordArray.map(({ letter, complete }, index) => (
                    <span key={index} className={cx('letter', { complete })}>
                        {letter}
                        <div style={{ transitionDuration }} className={styles.inner}>
                            {letter}
                        </div>
                    </span>
                ))}
            </div>
        );
    }
}

Word.propTypes = {
    wordArray: PropTypes.array, // todo proptypes
};

export default Word;
