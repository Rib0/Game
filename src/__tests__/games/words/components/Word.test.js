import React from 'react';
import Word from 'games/words/components/Word';

const setUp = props => shallow(<Word {...props} />);

describe('Words game - Word test', () => {
    let Component, instance, componentDidMount, componentWillUnmount;
    const props = {
        wordArray: [
            {
                letter: 'с',
                complete: false,
                missed: false
            },
            {
                letter: 'ы',
                complete: false,
                missed: false
            },
            {
                letter: 'р',
                complete: false,
                missed: false
            }
        ],
        getRightWord: jest.fn(),
        getRightLetter: jest.fn(),
        resetWord: jest.fn()
    };

    beforeEach(() => {
        Component = setUp(props);
        instance = Component.instance();
        componentDidMount = jest.spyOn(Word.prototype, 'componentDidMount');
        componentWillUnmount = jest.spyOn(Word.prototype, 'componentWillUnmount');
    });
    afterEach(() => {
        jest.useRealTimers();
        jest.restoreAllMocks();
        props.getRightWord.mockClear();
        props.getRightLetter.mockClear();
        props.resetWord.mockClear();
        componentDidMount.mockClear();
        componentWillUnmount.mockClear();
    });

    test('should render', () => {
        expect(Component).toMatchSnapshot();
    })
    test('AddEventListener should called once', () => {
        const addEventListener = jest.spyOn(document.body, 'addEventListener');
        instance.componentDidMount();
        expect(addEventListener).toHaveBeenCalledTimes(1);
    })
    test('RemoveEventListener should called once', () => {
        const removeEventListener = jest.spyOn(document.body, 'removeEventListener');
        Component.unmount();
        expect(removeEventListener).toHaveBeenCalledTimes(1);
    })
    test('ComponentDidMount should called once', () => {
        instance.componentDidMount();
        expect(componentDidMount).toHaveBeenCalledTimes(1);
    })
    test('componentWillUnmount should called once', () => {
        Component.unmount();
        expect(componentWillUnmount).toHaveBeenCalledTimes(1);
    })
    test('resetTransition should change transitionDuration', () => {
        jest.useFakeTimers('legacy');
        expect(Component.state().transitionDuration).toBe('.2s');
        instance.resetTransition();
        expect(Component.state().transitionDuration).toBe('0s');
        expect(setTimeout).toHaveBeenCalledTimes(1);
        jest.runAllTimers();
        expect(Component.state().transitionDuration).toBe('.2s');
    })
    test('complete word should trigger 2 methods', () => {
        const newProps = {
            ...props,
            wordArray: props.wordArray.map(word => ({ ...word, complete: true }))
        };
        const resetTransition = jest.spyOn(instance, 'resetTransition');
        Component.setProps(newProps);
        expect(newProps.getRightWord).toHaveBeenCalledTimes(1);
        expect(resetTransition).toHaveBeenCalledTimes(1);
    })
    test('should find index of missed letter', () => {
        jest.useFakeTimers('legacy');
        expect(Component.state().missedIndex).toBeNull();
        instance.handleMiss();
        expect(Component.state().missedIndex).toBe(0);
        jest.runAllTimers();
        expect(Component.state().missedIndex).toBeNull();
    })
    test('should handle keydown with right letter', () => {
        instance.onKeyDown({ key: 'с' });
        expect(props.getRightLetter).toHaveBeenCalledTimes(1)
    })
    test('should handle keydown with wrong letter', () => {
        const handleMiss = jest.spyOn(instance, 'handleMiss');
        const resetTransition = jest.spyOn(instance, 'resetTransition');
        instance.onKeyDown({ key: 'a' });
        expect(handleMiss).toHaveBeenCalledTimes(1);
        expect(props.resetWord).toHaveBeenCalledTimes(1);
        expect(resetTransition).toHaveBeenCalledTimes(1);
    })
})