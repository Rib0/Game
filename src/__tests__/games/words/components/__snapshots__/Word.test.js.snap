// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Words game - Word test should render 1`] = `
<div
  className="word"
>
  <span
    className="letter"
    key="0"
  >
    с
    <div
      className="inner"
      style={
        Object {
          "transitionDuration": ".2s",
        }
      }
    >
      с
    </div>
  </span>
  <span
    className="letter"
    key="1"
  >
    ы
    <div
      className="inner"
      style={
        Object {
          "transitionDuration": ".2s",
        }
      }
    >
      ы
    </div>
  </span>
  <span
    className="letter"
    key="2"
  >
    р
    <div
      className="inner"
      style={
        Object {
          "transitionDuration": ".2s",
        }
      }
    >
      р
    </div>
  </span>
</div>
`;
